samples/nrf9160/aws_fota/src/main.c

```diff
@@ -0,0 +1,584 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <uart.h>
+#include <string.h>
+
+#include <net/mqtt.h>
+#include <net/socket.h>
+#include <net/aws_fota.h>
+#include <lte_lc.h>
+#if defined(CONFIG_BSD_LIBRARY)
+
+#include "nrf_inbuilt_key.h"
+#endif
+
+#define NRF_CLOUD_SEC_TAG 16842753
+#define AWS_HOSTNAME "a25jld2wxwm7qs-ats.iot.eu-central-1.amazonaws.com"
+
+#include "aws_cert/rootca_ec.h"
+#include "aws_cert/private_key.h"
+#include "aws_cert/pubcert.h"
+
+#if !defined(CONFIG_CLIENT_ID)
+#define IMEI_LEN 15
+#define CLIENT_ID_LEN (IMEI_LEN + 4)
+#else
+#define CLIENT_ID_LEN (sizeof(CLOUD_CLIENT_ID) - 1)
+#endif
+static u8_t client_id_buf[CLIENT_ID_LEN+1];
+
+/* Buffers for MQTT client. */
+static u8_t rx_buffer[CONFIG_MQTT_MESSAGE_BUFFER_SIZE];
+static u8_t tx_buffer[CONFIG_MQTT_MESSAGE_BUFFER_SIZE];
+static u8_t payload_buf[CONFIG_MQTT_PAYLOAD_BUFFER_SIZE];
+
+/* MQTT Broker details. */
+static struct sockaddr_storage broker;
+
+/* File descriptor */
+static struct pollfd fds;
+
+/* Set to true when application should teardown and reboot */
+static bool do_reboot;
+
+#if defined(CONFIG_BSD_LIBRARY)
+
+/**@brief Recoverable BSD library error. */
+void bsd_recoverable_error_handler(uint32_t err)
+{
+	printk("bsdlib recoverable error: %u\n", err);
+}
+
+/**@brief Irrecoverable BSD library error. */
+void bsd_irrecoverable_error_handler(uint32_t err)
+{
+	printk("bsdlib irrecoverable error: %u\n", err);
+
+	__ASSERT_NO_MSG(false);
+}
+
+#endif /* defined(CONFIG_BSD_LIBRARY) */
+
+/**@brief Function to print strings without null-termination. */
+static void data_print(u8_t *prefix, u8_t *data, size_t len)
+{
+	char buf[len + 1];
+
+	memcpy(buf, data, len);
+	buf[len] = 0;
+	printk("%s%s\n", prefix, buf);
+}
+
+/**@brief Function to publish data on the configured topic. */
+static int data_publish(struct mqtt_client *c, enum mqtt_qos qos,
+			u8_t *data, size_t len)
+{
+	struct mqtt_publish_param param;
+
+	param.message.topic.qos = qos;
+	param.message.topic.topic.utf8 = CONFIG_MQTT_PUB_TOPIC;
+	param.message.topic.topic.size = strlen(CONFIG_MQTT_PUB_TOPIC);
+	param.message.payload.data = data;
+	param.message.payload.len = len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+	param.retain_flag = 0;
+
+	data_print("publishing: ", data, len);
+	return mqtt_publish(c, &param);
+}
+
+
+
+/**@brief Function to read the published payload. */
+static int publish_get_payload(struct mqtt_client *c, size_t length)
```

Comment:
You can copy the version from `aws_fota` lib here, which was more thoroughly reviewed and is simpler.

Reply:
Done